src/peft/tuners/tuners_utils.py

```diff
@@ -361,7 +361,33 @@ def _check_merge_allowed(self):
 
         Raise a ValueError if it is not possible to merge the adapter with the given configuration.
         """
-        pass
+        tied_target_modules = self._get_tied_target_modules(self.model)
+        if tied_target_modules:
+            warnings.warn(
+                f"Model with `tie_word_embeddings=True` and the {tied_target_modules=} are part of the adapter. "
+                "This can lead to complications. "
+                "You can opt to merge the adapter after cloning the weights (to untie the embeddings). "
+                "You can untie the embeddings by loading the model with `tie_word_embeddings=False`. For example:"
+                """
+```python
+from transformers import AutoModelForCausalLM
+
+# Load original tied model
+model = AutoModelForCausalLM.from_pretrained("google/gemma-2-2b-it", tie_word_embeddings=False)
+
+# Set the randomly initialized lm_head to the previously tied embeddings
+model.lm_head.weight.data = model.model.embed_tokens.weight.data.clone()
+
+# Save the untied model
+untied_model_dir = "dir/for/untied/model"
+model.save_pretrained(untied_model_dir)
+model.config.save_pretrained(untied_model_dir)
+
+# Now use the original model but in untied format
+model = AutoModelForCausalLM.from_pretrained(untied_model_dir)
+```
+"""
```

Comment:
I see why you left-aligned the code snippet so that it is nicely printed. But this is really an eye-sore to read in code. Here is a trick to that let's us use the correct indentation but still get a nice warning message by using `textwrap.dedent`:

```python
            example_code = textwrap.dedent(
                """
                ```python
                from transformers import AutoModelForCausalLM

                # Load original tied model
                model = AutoModelForCausalLM.from_pretrained("google/gemma-2-2b-it", tie_word_embeddings=False)

                # Set the randomly initialized lm_head to the previously tied embeddings
                model.lm_head.weight.data = model.model.embed_tokens.weight.data.clone()

                # Save the untied model
                untied_model_dir = "dir/for/untied/model"
                model.save_pretrained(untied_model_dir)
                model.config.save_pretrained(untied_model_dir)

                # Now use the original model but in untied format
                model = AutoModelForCausalLM.from_pretrained(untied_model_dir)
                ```
                """
            )
            warnings.warn(
                f"Model with `tie_word_embeddings=True` and the {tied_target_modules=} are part of the adapter. "
                "This can lead to complications. "
                "You can opt to merge the adapter after cloning the weights (to untie the embeddings). "
                "You can untie the embeddings by loading the model with `tie_word_embeddings=False`. For example:"
                + example_code
            )
```

The `textwrap` module is from the standardlib and needs to be imported.

Reply:
Addressed both thanks!